### PR TITLE
Monomorphize method applications

### LINKED ISSUE #1628: First use of a generic method fixes its generic type
Here, the first use of `fn write<T>(self, val: T, offset: u64)` with a `bool` sets its `T` to `bool` and prevents it from being called with a `u64` (or something else) later.
```rs
script;

use std::mem::*;
use std::intrinsics::*;
use std::assert::assert;

struct TestStruct {
    boo: bool,
    uwu: u64
}

fn main() -> bool {
    // Create a struct
    let foo = TestStruct { boo: true, uwu: 42 };
    let foo_len = size_of::<TestStruct>();
    assert(foo_len == 16);

    // Create a clone of the struct
    let buf = ~Buffer::alloc(foo_len);
    buf.write(true, 0);
    buf.write(42, size_of::<bool>());
//            ^^ This parameter was declared as type bool, but argument of type u64 was provided.
    let foo: TestStruct = buf.into_unchecked();
    assert(foo.boo == true);
    assert(foo.uwu == 42);

    true
}
```

The `Buffer` in question:
```rs
pub struct Buffer {
    ptr: Pointer,
    len: u64
}

impl Buffer {
    // ... snip ...

    /// Writes the given value into the buffer
    pub fn write<T>(self, val: T, offset: u64) {
        if is_reference_type::<T>() {
            let ptr = ~Pointer::from(val);
            let len = size_of::<T>();
            
            // We can't reference `self.write_unchecked` like this:
            // self.write_unchecked(ptr, len, offset);

            // Instead we inline
            copy(self.ptr.with_offset(offset), ptr, len);
        } else {
            let dst_ptr = self.ptr.with_offset(offset);
            asm(ptr: dst_ptr.val, val: val) {
                sw ptr val i0;
            };
        }
    }

    // ... snip ...
}
```
You can find the full code of `Buffer` (and `std::mem`) here: https://github.com/FuelLabs/sway/blob/522bf95b7f06e679161712ae4a0821c74f318a2c/sway-lib-std/src/mem.sw

### PR DESCRIPTION
**Pull request by sezna**
We were not monomorphizing method applications at all. This PR fixes that. This change was made much easier by #1589  and #1666 

Closes #1628 

This PR works by monomorphizing declarations when methods are instantiated against them.  This now means that you can annotate types for a parent type and a function, like Rust.

So, this example works:

```rust

struct Foo<T> {
  a: T
}
impl <T> Foo {
   fn bar<O>(o: O) -> u64 {
     5
  }
}
fn main() {
      // note the annotation `u64` for `Foo` and `bool` for `bar()`
    Foo::<u64>::bar::<bool>(true)
}

```

This involved changes to the parser to allow generic arguments in the new position and calling `monomorphize` on declarations when methods are instantiated. All the other changes are just fallout from those two things.

edit: some interaction with recent PRs has broken some tests, moving to draft while that gets sorted out...

### CODE REVIEW

**Review by AlicanC — COMMENTED** [review]
I tried this on my local while working on the Multicall script and it indeed solves my issues! Can do things a lot more conveniently now 👍 👍 👍 👍

File: sway-core/src/convert_parse_tree.rs [hunk 1]
```diff
@@ -1238,53 +1238,60 @@ fn expr_to_expression(ec: &mut ErrorContext, expr: Expr) -> Result<Expression, E
...
+                match suffix.pop() {
+                    Some((_double_colon_token, call_path_suffix)) => match suffix.pop() {
+                        Some((_double_colon_token, maybe_method_segment)) => {
+                            let PathExprSegment {
+                                fully_qualified,
+                                name,
+                                generics_opt,
+                            } = maybe_method_segment;
+                            if let Some((_double_colon_token, generic_args)) = generics_opt {
+                                println!("here?1");
```
Comment (otrho) [inline sway-core/src/convert_parse_tree.rs:1251]:
```suggestion
```

**Review by otrho — CHANGES_REQUESTED** [review]
LGTM, though I'm not the type checker expert.

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression/method_application.rs [hunk 2]
```diff
@@ -63,6 +64,21 @@ pub(crate) fn type_check_method_application(
         None
     };
 
+    // monomorphize the parent type
```
Comment (emilyaherbert) [inline sway-core/src/semantic_analysis/ast_node/expression/typed_expression/method_application.rs:67]: remove

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression/method_application.rs [hunk 3]
```diff
@@ -375,10 +391,64 @@ fn find_method(
         }
         (ty, true) => ty.clone(),
     };
+    let ty = (|| -> CompileResult<TypeId> {
```
Comment (emilyaherbert) [inline sway-core/src/semantic_analysis/ast_node/expression/typed_expression/method_application.rs:394]: Is it necessary to use a function closure here? This makes it harder to read and maintain IMO.

File: test/src/e2e_vm_tests/test_programs/should_pass/language/generic_impl_self/src/main.sw [hunk 4]
```diff
@@ -54,7 +54,7 @@ impl DoubleIdentity<u8, u8> {
 }
 
 fn double_identity2<T, F>(x: T, y: F) -> DoubleIdentity<T, F> {
-  ~DoubleIdentity::new::<T, F>(x, y)
+  ~DoubleIdentity::<T, F>::new(x, y)
```
Comment (emilyaherbert) [inline test/src/e2e_vm_tests/test_programs/should_pass/language/generic_impl_self/src/main.sw:57]: Can you explain the reason for changing the syntax?

**Review by emilyaherbert — CHANGES_REQUESTED** [review]

**Comment by emilyaherbert** [comment]
Based on our convo @sezna, I'm going to close this PR, copy your parser changes to a new PR, then take over https://github.com/FuelLabs/sway/issues/1628.

**Pull request closed without merging**